### PR TITLE
Update ray to 2.7.0, update patch

### PR DESCRIPTION
**Pull request by timkpaine**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/patches/0005-Remove-all-dependencies-from-setup.py.patch [hunk 1]
```diff
@@ -53,8 +53,15 @@ index affa582097..a26f61f8fb 100644
...
 -        ],
--        "serve": ["uvicorn", "requests", "starlette", "fastapi", "aiorwlock"],
--        "tune": ["pandas", "tensorboardX>=1.9", "requests", pyarrow_dep],
+-        "serve": [
+-            "uvicorn",
+-            "requests",
+-            "starlette",
+-            "fastapi",
+-            "aiorwlock",
+-            "watchfiles",
```
Comment (mattip) [inline recipe/patches/0005-Remove-all-dependencies-from-setup.py.patch:62]: `watchfiles` is a new dependency for ray-serve

File: recipe/patches/0005-Remove-all-dependencies-from-setup.py.patch [hunk 2]
```diff
@@ -53,8 +53,15 @@ index affa582097..a26f61f8fb 100644
...
+-        ],
+-        "tune": ["pandas", "tensorboardX>=1.9", "requests", pyarrow_dep, "fsspec"],
```
Comment (mattip) [inline recipe/patches/0005-Remove-all-dependencies-from-setup.py.patch:64]: `fsspec` is a new dependency for ray-tune

File: recipe/patches/0005-Remove-all-dependencies-from-setup.py.patch [hunk 3]
```diff
@@ -66,19 +73,34 @@ index affa582097..a26f61f8fb 100644
...
+-        )
+-    )
 -
      if RAY_EXTRA_CPP:
          setup_spec.extras["cpp"] = ["ray-cpp==" + setup_spec.version]
  
 -    setup_spec.extras["rllib"] = setup_spec.extras["tune"] + [
 -        "dm_tree",
--        "gymnasium==0.26.3",
+-        "gymnasium==0.28.1",
```
Comment (mattip) [inline recipe/patches/0005-Remove-all-dependencies-from-setup.py.patch:103]: The gymnasium version has changed for rllib

**Comment by mattip** [comment]
ray-dashboard test is failing. I think it needs to depend on ray-default, not ray-core.
```
+ python -c 'import ray; ray.init(include_dashboard=True)'
2023-09-25 21:21:18,285	ERROR services.py:1129 -- --include-dashboard is not supported when minimal ray is used.Download ray[default] to use the dashboard.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695666789674/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/_private/client_mode_hook.py", line 103, in wrapper
    return func(*args, **kwargs)
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695666789674/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/_private/worker.py", line 1536, in init
    _global_node = ray._private.node.Node(
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695666789674/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/_private/node.py", line 307, in __init__
    self.start_head_processes()
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695666789674/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/_private/node.py", line 1419, in start_head_processes
    self.start_api_server(
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695666789674/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/_private/node.py", line 1168, in start_api_server
    self._webui_url, process_info = ray._private.services.start_api_server(
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695666789674/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/_private/services.py", line 1289, in start_api_server
    raise e from e
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695666789674/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/_private/services.py", line 1133, in start_api_server
    raise Exception("Cannot include dashboard with missing packages.")

```

**Comment by mattip** [comment]
Somehow the dependencies must be reworked
- ray-dashboard needs to depend on ray-default
- whatever is causing the cycle needs to be broken.

**Comment by mattip** [comment]
Hmm. linux-python3.10 is failing. Something about protobuf wanting libc++ with a different version of CXXABI. @h-vetinari have you seen that before?
```
import: 'ray.dashboard'
+ python -c 'import ray; ray.init(include_dashboard=True)'
2023-09-27 13:19:52,730	ERROR services.py:1222 -- Failed to start the dashboard , return code 1
2023-09-27 13:19:52,731	ERROR services.py:1247 -- Error should be written to 'dashboard.log' or 'dashboard.err'. We are printing the last 20 lines for you. See 'https://docs.ray.io/en/master/ray-observability/ray-logging.html#logging-directory-structure' to find where the log file is.
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695814102518/_test_env_.../lib/python3.10/site-packages/ray/_private/node.py", line 1168, in start_api_server
    self._webui_url, process_info = ray._private.services.start_api_server(
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695814102518/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/_private/services.py", line 1289, in start_api_server
    raise e from e
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695814102518/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/_private/services.py", line 1277, in start_api_server
    raise Exception(last_log_str)
Exception: 
The last 20 lines of /tmp/ray/session_2023-09-27_13-19-51_338636_52832/logs/dashboard.err (it contains the error message from the dashboard): 
  warnings.warn(
Traceback (most recent call last):
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695814102518/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/dashboard/dashboard.py", line 10, in <module>
    import ray._private.ray_constants as ray_constants
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695814102518/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/__init__.py", line 96, in <module>
    import ray._raylet  # noqa: E402
  File "python/ray/_raylet.pyx", line 165, in init ray._raylet
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695814102518/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/exceptions.py", line 10, in <module>
    from ray.core.generated.common_pb2 import (
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695814102518/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/ray/core/generated/common_pb2.py", line 6, in <module>
    from google.protobuf import descriptor as _descriptor
  File "/home/conda/feedstock_root/build_artifacts/ray-packages_1695814102518/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pla/lib/python3.10/site-packages/google/protobuf/descriptor.py", line 51, in <module>
    from google.protobuf.pyext import _message
ImportError: /lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by /home/conda/feedstock_root/build_artifacts/ray-packages_1695814102518/_test_env_.../lib/python3.10/site-packages/google/protobuf/pyext/_message.cpython-310-x86_64-linux-gnu.so)


```

**Comment by mattip** [comment]
@isuruf have you seen this before?
```
   from google.protobuf.pyext import _message
ImportError: /lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found \
    (required by /home/conda/feedstock_root/build_artifacts/ray-packages_1695814102518 \
    /_test_env_.../lib/python3.10/site-packages/google/protobuf/pyext/_message.cpython-310-x86_64-linux-gnu.so)
```

**Comment by isuruf** [comment]
> `ERROR (ray-core,lib/python3.10/site-packages/ray/core/libjemalloc.so): $PATH/lib/libstdc++.so.6 found in build prefix; should never happen`

`libjemalloc.so` is not using the correct RPATH.

**Comment by isuruf** [comment]
You need to add `ray/core/jemalloc.so` to https://github.com/conda-forge/ray-packages-feedstock/blob/2f76deecabd49346a59b372c940f41407fe21cb0/recipe/build-core.sh#L25

**Comment by mattip** [comment]
The gcs server is failing to start. Replicating locally, I see this in gcs_server.err
```
*** SIGSEGV received at time=1696401255 on cpu 19 ***
PC: @     0x55d8b7dc9d72  (unknown)  std::codecvt<>::do_unshift()
    @     0x55d8b7d6a54d         64  absl::lts_20220623::AbslFailureSignalHandler()
    @     0x7fecbab35520  1190386256  (unknown)
    @     0x7fecba4e2000  (unknown)  (unknown)
[failure_signal_handler.cc : 329] RAW: Signal 11 raised at PC=0x55d8b7dc9d72 while already in AbslFailureSignalHandler()
*** SIGSEGV received at time=1696401255 on cpu 19 ***
PC: @     0x55d8b7dc9d72  (unknown)  std::codecvt<>::do_unshift()
    @     0x55d8b7d6a54d         64  absl::lts_20220623::AbslFailureSignalHandler()
    @     0x7fecbab35520       1472  (unknown)
    @     0x7fecba4e21c0  (unknown)  (unknown)
[
```

I am trying to track this down, and hints are welcome.

**Comment by mattip** [comment]
This is when testing ray-core, which means there is no conda/pip installed grpc. The 2.7 build now links to a static build of grpc pulled via the bazel build, [version 1.50.2?](https://github.com/ray-project/ray/blob/67593a914a7d2e5dd703a27bad9959a1ceb94bf9/bazel/ray_deps_setup.bzl#L243) and does not need a python grpc package for ray-core.

**Comment by mattip** [comment]
RAy uses [`LD_PRELOAD` to force use of `libjemalloc`](). I can reproduce the segfault if I do the same:
```
LD_PRELOAD=<path>/lib/python3.9/site-packages/ray/core/libjemalloc.so \
    <path>/lib/python3.9/site-packages/ray/core/src/ray/gcs/gcs_server \
    --log_dir=/tmp/logs --config_list=<hash> --gcs_server_port=45547 \
    --metrics-agent-port=61490 --node-ip-address=10.0.0.19 --session-name=<session-name>
```

The `gcs_server` process runs without crashing if I do not use `libjemalloc.so`.

**Comment by mattip** [comment]
Here is the backtrace from gdb
```
gdb) bt
#0  std::codecvt<char16_t, char, __mbstate_t>::do_unshift (this=0x555556650e80 <(anonymous namespace)::codecvt_c16>, 
    __to=0x5555561b8000 <std::__ostream_insert<char, std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*, long)+294> "\353\a\363\017\036\372H\211\303H\211\357\350o\366\377\377H\211\337\350wKM\377H\211\357\350_\366\377\377H\203\304\020H\211\330[]A\\A]A^\303\363\017\036\372SH\211\373H\203\354\020H\205\366u\a\350\n\365\377\377\353\035H\211\367H\211t$\b\350\212KM\377H\213t$\bH\211\337H\211\302\350t\376\377\377H\203\304\020H\211\330[Ð\363\017\036\372H\213\006H\213@\350H\213\204", <incomplete sequence \350>, __to_next=<error reading variable: Cannot access memory at address 0x20>)
    at ../../../../../libstdc++-v3/src/c++11/codecvt.cc:769
#1  0x00005555561b8114 in std::ostream::_M_insert<long> (this=0x7ffff72e2000, __v=35)
    at /home/conda/feedstock_root/build_artifacts/gcc_compilers_1695216551398/work/build/x86_64-conda-linux-gnu/libstdc++-v3/include/bits/ostream.tcc:78
#2  0x00005555560295ee in ray::RayLog::RayLog(char const*, int, ray::RayLogLevel) ()
#3  0x0000555555b4a9ea in ray::IOServicePool::Run() ()
#4  0x00005555556f3f27 in main ()
```

**Comment by mattip** [comment]
More strangeness: If I modify the code to use `libjemalloc.so` from pip or from conda, or the one built in this package, I always get a segfault, even though I can successfully run `LD_PRELOAD=<path>/libjemalloc.so gcs_server ...` from the CLI. Ther must be some global symbol already loaded in the executable that does not play well with `libjemalloc`.

**Comment by mattip** [comment]
Digging through the `Config.log` for libjemalloc built in this feedstock, I see it uses `-static-libstdc++`, where the one built by a git checkout (that works properly) does not. That seems to be the only significant difference in the `Config.log` between the two.

**Comment by mattip** [comment]
CI is passing. I will remove the "draft" label. Could you remove the stray grpcio dependency from here? It seems to be leftover from a previous iteration. https://github.com/conda-forge/ray-packages-feedstock/blob/1d075591177d1188bfa0e6790719ca8aa546be02/recipe/meta.yaml#L202-L204

**Comment by mattip** [comment]
Sorry, not remove, rather make it `grpcio  #[not osx]`

**Comment by mattip** [comment]
@h-vetinari once this is in, we should revisit #90, #98

File: recipe/patches/0005-Remove-all-dependencies-from-setup.py.patch [hunk 4]
```diff
@@ -1,19 +1,17 @@
...
-From: Kai Fricke <kai@anyscale.com>
-Date: Fri, 3 Dec 2021 10:55:23 -0800
-Subject: [PATCH 5/5] Remove all dependencies from setup.py
+From d15934405a7f9f5234e2b612dd158610c4fa0261 Mon Sep 17 00:00:00 2001
+From: Tim Paine <3105306+timkpaine@users.noreply.github.com>
+Date: Mon, 25 Sep 2023 18:29:15 +0000
+Subject: [PATCH] Remove all dependencies from setup.py
 
-Signed-off-by: Jiajun Yao <jeromeyjj@gmail.com>
-Signed-off-by: Matti Picus <matti.picus@gmail.com>
```
Comment (h-vetinari) [inline recipe/patches/0005-Remove-all-dependencies-from-setup.py.patch:7]: This should not change, though you can add your `Signed-off-by`.

Also, the patch should ideally be generated with `git format-patch --no-signature`

**Comment by mattip** [comment]
Merging, even though there is 2.7.1

**Pull request merged**